trinity/protocol/common/monitors.py

```diff
@@ -0,0 +1,93 @@
+import asyncio
+from contextlib import contextmanager
+from typing import (
+    AsyncIterator,
+    Iterator,
+    Set,
+)
+
+from cancel_token import CancelToken
+from eth_utils import ValidationError
+
+from p2p.exceptions import NoConnectedPeers
+from p2p.peer import BasePeer, PeerSubscriber
+from p2p.service import BaseService
+
+from trinity.protocol.common.peer import BaseChainPeer, BaseChainPeerPool
+from trinity.protocol.eth import commands
+
+
+class BaseChainTipMonitor(BaseService, PeerSubscriber):
+    """
+    Monitor for potential changes to the tip of the chain: a new peer or a new block
+
+    Subclass must specify :attr:`subscription_msg_types`
+    """
+
+    def __init__(
+            self,
+            peer_pool: BaseChainPeerPool,
+            token: CancelToken = None) -> None:
+        super().__init__(token)
+        self._peer_pool = peer_pool
+        # There is one event for each subscriber, each one gets set any time new tip info arrives
+        self._subscriber_notices: Set[asyncio.Event] = set()
+
+    async def wait_tip_info(self) -> AsyncIterator[BaseChainPeer]:
+        """
+        This iterator waits until there is potentially new tip information.
+        New tip information means a new peer connected or a new block arrived.
+        Then it yields the peer with the highest total difficulty.
+        It continues indefinitely, until this service is cancelled.
+        """
+        if self.is_cancelled:
+            raise ValidationError("%s is cancelled, new tip info is impossible", self)
+        elif not self.is_running:
+            await self.events.started.wait()
+
+        with self._subscriber() as new_tip_event:
+            while self.is_operational:
+                try:
+                    highest_td_peer = self._peer_pool.highest_td_peer
+                except NoConnectedPeers:
+                    # no peers are available right now, skip the new tip info yield
```

Comment:
I think this might be where your problem is.  This loop looks like it won't ever yield the process if this `except` clause is hit.  I think you need to inject an `await self.sleep(0)` here somewhere to ensure that you yield the event loop at some point.

Reply:
Yup, looks right to me, thanks!